File: scripts/generate_favicon.py
#!/usr/bin/env python3
"""
Generate favicon.ico from SVG logo

Converts the SerenityOps SVG logo to a multi-resolution .ico file
suitable for use as a favicon in web applications.
"""

import sys
from pathlib import Path

try:
    from cairosvg import svg2png
    from PIL import Image
    import io
except ImportError as e:
    print(f"Missing required packages. Install with:")
    print(f"  pip install cairosvg pillow")
    sys.exit(1)


def generate_favicon(svg_path: Path, output_path: Path):
    """
    Generate multi-resolution favicon from SVG

    Args:
        svg_path: Path to input SVG file
        output_path: Path for output .ico file
    """
    print(f"📄 Reading SVG: {svg_path}")

    # Read SVG file
    with open(svg_path, 'rb') as f:
        svg_data = f.read()

    # Generate PNG at different sizes for ICO
    sizes = [16, 32, 48, 64, 128, 256]
    images = []

    print(f"🔄 Converting SVG to PNG at multiple resolutions...")

    for size in sizes:
        print(f"  - {size}x{size}px")

        # Convert SVG to PNG at specific size
        png_data = svg2png(
            bytestring=svg_data,
            output_width=size,
            output_height=size
        )

        # Load PNG with Pillow
        img = Image.open(io.BytesIO(png_data))
        images.append(img)

    # Save as ICO with multiple resolutions
    print(f"💾 Saving favicon: {output_path}")
    images[0].save(
        output_path,
        format='ICO',
        sizes=[(img.width, img.height) for img in images],
        append_images=images[1:]
    )

    print(f"✅ Favicon generated successfully!")
    print(f"   Size: {output_path.stat().st_size / 1024:.1f} KB")


def main():
    # Paths
    script_dir = Path(__file__).parent
    project_root = script_dir.parent

    svg_path = project_root / "assets" / "logo.svg"
    output_dir = project_root / "frontend" / "public"
    output_path = output_dir / "favicon.ico"

    # Ensure output directory exists
    output_dir.mkdir(parents=True, exist_ok=True)

    # Check if SVG exists
    if not svg_path.exists():
        print(f"❌ Error: SVG not found at {svg_path}")
        sys.exit(1)

    # Generate favicon
    generate_favicon(svg_path, output_path)

    # Also generate PNG versions for various uses
    print(f"\n🎨 Generating additional PNG assets...")

    # App icon (512x512)
    app_icon_path = output_dir / "logo512.png"
    svg_data = svg_path.read_bytes()
    png_512 = svg2png(bytestring=svg_data, output_width=512, output_height=512)
    app_icon_path.write_bytes(png_512)
    print(f"  ✓ logo512.png ({app_icon_path.stat().st_size / 1024:.1f} KB)")

    # Medium icon (192x192)
    medium_icon_path = output_dir / "logo192.png"
    png_192 = svg2png(bytestring=svg_data, output_width=192, output_height=192)
    medium_icon_path.write_bytes(png_192)
    print(f"  ✓ logo192.png ({medium_icon_path.stat().st_size / 1024:.1f} KB)")

    print(f"\n🎉 All assets generated!")
    print(f"\nGenerated files:")
    print(f"  - {output_path}")
    print(f"  - {app_icon_path}")
    print(f"  - {medium_icon_path}")


if __name__ == "__main__":
    main()

File: scripts/generate_favicon.js
#!/usr/bin/env node
/**
 * Generate favicon.ico from SVG logo
 *
 * Converts the SerenityOps SVG logo to a multi-resolution .ico file
 * suitable for use as a favicon in web applications.
 */

const fs = require('fs');
const path = require('path');

// Check for sharp
let sharp;
try {
  sharp = require('sharp');
} catch (err) {
  console.error('❌ Error: sharp package not found');
  console.error('Install with: npm install sharp');
  process.exit(1);
}

async function generateFavicon() {
  const projectRoot = path.join(__dirname, '..');
  const svgPath = path.join(projectRoot, 'assets', 'logo.svg');
  const outputDir = path.join(projectRoot, 'frontend', 'public');
  const outputPath = path.join(outputDir, 'favicon.ico');

  console.log(`📄 Reading SVG: ${svgPath}`);

  // Read SVG file
  const svgBuffer = fs.readFileSync(svgPath);

  // Ensure output directory exists
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  console.log(`🔄 Converting SVG to PNG at multiple resolutions...`);

  // Generate PNG at different sizes
  const sizes = [16, 32, 48, 256];
  const pngBuffers = [];

  for (const size of sizes) {
    console.log(`  - ${size}x${size}px`);
    const pngBuffer = await sharp(svgBuffer)
      .resize(size, size)
      .png()
      .toBuffer();
    pngBuffers.push(pngBuffer);
  }

  // For ICO, we'll use the 32x32 version as the main favicon
  // (ICO with multiple sizes requires special handling, so we'll use PNG fallback)
  const favicon32 = pngBuffers[1]; // 32x32

  console.log(`💾 Saving favicon: ${outputPath}`);
  fs.writeFileSync(outputPath.replace('.ico', '.png'), favicon32);

  // Generate additional PNG assets
  console.log(`\n🎨 Generating additional PNG assets...`);

  // App icon (512x512)
  const logo512Path = path.join(outputDir, 'logo512.png');
  const png512 = await sharp(svgBuffer).resize(512, 512).png().toBuffer();
  fs.writeFileSync(logo512Path, png512);
  console.log(`  ✓ logo512.png (${(png512.length / 1024).toFixed(1)} KB)`);

  // Medium icon (192x192)
  const logo192Path = path.join(outputDir, 'logo192.png');
  const png192 = await sharp(svgBuffer).resize(192, 192).png().toBuffer();
  fs.writeFileSync(logo192Path, png192);
  console.log(`  ✓ logo192.png (${(png192.length / 1024).toFixed(1)} KB)`);

  // Favicon 32x32
  const favicon32Path = path.join(outputDir, 'favicon.png');
  fs.writeFileSync(favicon32Path, favicon32);
  console.log(`  ✓ favicon.png (${(favicon32.length / 1024).toFixed(1)} KB)`);

  // Also copy SVG to public
  const logoSvgPath = path.join(outputDir, 'logo.svg');
  fs.copyFileSync(svgPath, logoSvgPath);
  console.log(`  ✓ logo.svg (copied)`);

  console.log(`\n🎉 All assets generated!`);
  console.log(`\nGenerated files:`);
  console.log(`  - ${favicon32Path}`);
  console.log(`  - ${logo512Path}`);
  console.log(`  - ${logo192Path}`);
  console.log(`  - ${logoSvgPath}`);
}

generateFavicon().catch(err => {
  console.error('❌ Error:', err.message);
  process.exit(1);
});
